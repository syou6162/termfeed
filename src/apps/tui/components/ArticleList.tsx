import { Box, Text, useStdout } from 'ink';
import { useState, useEffect, memo, useMemo } from 'react';
import type { Article } from '../../../types/index.js';
import { convertHtmlToText } from '../utils/html.js';
import { TUI_CONFIG } from '../config/constants.js';

type ArticleListProps = {
  articles: Article[];
  selectedArticle?: Article;
  scrollOffset: number;
  onScrollOffsetChange: (offset: number) => void;
  isPinned?: boolean;
};

export const ArticleList = memo(function ArticleList({
  articles,
  selectedArticle,
  scrollOffset,
  onScrollOffsetChange,
  isPinned = false,
}: ArticleListProps) {
  const { stdout } = useStdout();
  const [contentLines, setContentLines] = useState<string[]>([]);
  const [totalLines, setTotalLines] = useState(0);

  // 記事が変わったときにスクロール位置をリセット
  useEffect(() => {
    if (selectedArticle) {
      onScrollOffsetChange(0);
    }
  }, [selectedArticle?.id, onScrollOffsetChange]);

  // コンテンツの行を更新
  useEffect(() => {
    if (selectedArticle?.content) {
      const plainText = convertHtmlToText(selectedArticle.content);
      const lines = plainText.split('\n');
      setContentLines(lines);
      setTotalLines(lines.length);
    } else {
      setContentLines([]);
      setTotalLines(0);
    }
  }, [selectedArticle?.content]);

  // スクロールオフセットを最大値に制限
  useEffect(() => {
    const totalHeight = stdout?.rows || TUI_CONFIG.DEFAULT_TERMINAL_HEIGHT;
    const availableLines = Math.max(1, totalHeight - TUI_CONFIG.ARTICLE_FIXED_LINES);
    const maxOffset = Math.max(0, totalLines - availableLines);

    if (scrollOffset > maxOffset) {
      onScrollOffsetChange(maxOffset);
    }
  }, [scrollOffset, totalLines, stdout?.rows, onScrollOffsetChange]);

  // ターミナルの高さから表示可能な行数を計算
  const totalHeight = stdout?.rows || TUI_CONFIG.DEFAULT_TERMINAL_HEIGHT;
  const availableLines = Math.max(1, totalHeight - TUI_CONFIG.ARTICLE_FIXED_LINES);

  // articlesは既に未読記事のみ
  const unreadCount = articles.length;
  const currentUnreadIndex = articles.findIndex((article) => article.id === selectedArticle?.id);
  const unreadPosition = currentUnreadIndex !== -1 ? currentUnreadIndex + 1 : 0;

  // useMemoは条件分岐の前に配置（Reactフックのルール）
  const publishedDate = useMemo(
    () =>
      selectedArticle?.published_at
        ? selectedArticle.published_at.toLocaleDateString('ja-JP', {
            year: 'numeric',
            month: 'long',
            day: 'numeric',
            hour: '2-digit',
            minute: '2-digit',
          })
        : '',
    [selectedArticle?.published_at]
  );

  if (articles.length === 0) {
    return (
      <Box flexDirection="column" height="100%" width="100%" padding={1} flexShrink={0}>
        <Box justifyContent="center" alignItems="center" height="95%" flexGrow={1}>
          <Box flexDirection="column" alignItems="center">
            <Text color="gray" italic>
              未読記事がありません
            </Text>
            <Text color="yellow">ヒント: `r` でフィードを更新できます</Text>
          </Box>
        </Box>
        <Box marginTop={1} justifyContent="center">
          <Text color="gray" dimColor>
            ? でヘルプ表示
          </Text>
        </Box>
      </Box>
    );
  }

  // 記事詳細を直接レンダリング（renderArticleDetail関数を使わない）
  if (!selectedArticle) {
    return (
      <Box
        flexDirection="column"
        height="100%"
        width="100%"
        borderStyle="single"
        borderLeft
        flexShrink={0}
      >
        <Box flexGrow={1} padding={1} justifyContent="center" alignItems="center">
          <Text color="gray" italic>
            記事を選択してください
          </Text>
        </Box>
        <Box padding={1} flexDirection="row">
          <Box flexGrow={1}>
            <Text color="gray" dimColor>
              ? でヘルプ表示
            </Text>
          </Box>
          <Box>
            <Text color="gray">
              {unreadPosition > 0 ? `${unreadPosition}/${unreadCount}件` : `未読${unreadCount}件`}
            </Text>
          </Box>
        </Box>
      </Box>
    );
  }

  const content = selectedArticle.content || '';

  // スクロール位置に基づいて表示する行を選択
  const visibleLines = contentLines.slice(scrollOffset, scrollOffset + availableLines);
  const displayText = visibleLines.join('\n');

  // スクロール情報
  const hasMoreContent = scrollOffset + availableLines < totalLines;
  const scrollInfo =
    totalLines > availableLines
      ? `(${scrollOffset + 1}-${Math.min(scrollOffset + availableLines, totalLines)} / ${totalLines}行)`
      : '';

  return (
    <Box
      flexDirection="column"
      borderStyle="single"
      borderLeft
      height="100%"
      width="100%"
      flexShrink={0}
    >
      {/* ヘッダー部分：固定 */}
      <Box paddingTop={1} paddingX={1}>
        <Text bold color="green">
          {selectedArticle.title}
        </Text>
      </Box>
      <Box paddingX={1}>
        <Text color="gray">公開日: {publishedDate}</Text>
        {selectedArticle.author && <Text color="cyan"> | 著者: {selectedArticle.author}</Text>}
        {selectedArticle.is_favorite && (
          <Text color="yellow" bold>
            {selectedArticle.author ? ' | ' : ' | '}★ お気に入り
          </Text>
        )}
        {isPinned && (
          <Text color="yellow" bold>
            {selectedArticle.author || selectedArticle.is_favorite ? ' | ' : ' | '}📌 ピン
          </Text>
        )}
      </Box>
      <Box paddingX={1} marginBottom={1}>
        <Text color="yellow">URL: {selectedArticle.url}</Text>
      </Box>

      {/* コンテンツ部分：スクロール可能 */}
      {content && (
        <Box paddingX={1} height={availableLines} overflow="hidden">
          <Text wrap="wrap">{displayText}</Text>
        </Box>
      )}

      {/* ステータス部分：固定 */}
      <Box paddingX={1} marginTop={1}>
        <Text color="gray" dimColor>
          {scrollInfo}
          {hasMoreContent && (scrollInfo ? ' ' : '') + 'スペースで続きを表示'}
        </Text>
      </Box>

      {/* フッター部分：固定 */}
      <Box padding={1} flexDirection="row">
        <Box flexGrow={1}>
          <Text color="gray" dimColor>
            ? でヘルプ表示
          </Text>
        </Box>
        <Box>
          <Text color="gray">
            {unreadPosition > 0 ? `${unreadPosition}/${unreadCount}件` : `未読${unreadCount}件`}
          </Text>
        </Box>
      </Box>
    </Box>
  );
});
