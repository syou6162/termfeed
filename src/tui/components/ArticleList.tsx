import { Box, Text } from 'ink';
import type { Article } from '../../models/types.js';
import { formatSummary } from '../utils/html.js';
import { SelectableList } from './SelectableList.js';

type ArticleListItem = {
  id: number;
  displayText: string;
  badge?: string;
  isRead?: boolean;
  isFavorite?: boolean;
};

type ArticleListProps = {
  articles: Article[];
  selectedIndex: number;
  selectedArticle?: Article;
  feedTitle?: string;
};

export function ArticleList({ articles, selectedIndex, selectedArticle, feedTitle }: ArticleListProps) {
  const articleItems: ArticleListItem[] = articles.map((article) => ({
    id: article.id || 0,
    displayText: article.title,
    isRead: article.is_read,
    isFavorite: article.is_favorite,
  }));

  const renderArticleItem = (item: ArticleListItem, isSelected: boolean) => {
    const prefix = isSelected ? '>' : ' ';
    const favoriteMarker = item.isFavorite ? '★ ' : '';
    const readMarker = item.isRead === false ? '● ' : '';
    const selectedStyle = isSelected ? { color: 'blue', bold: true } : {};

    return (
      <Text key={item.id} {...selectedStyle}>
        {prefix} {favoriteMarker}
        {readMarker}
        {item.displayText}
      </Text>
    );
  };

  const renderArticleDetail = () => {
    if (!selectedArticle) {
      return (
        <Box>
          <Text color="gray" italic>
            記事を選択してください
          </Text>
        </Box>
      );
    }

    const summary = formatSummary(selectedArticle.summary || selectedArticle.content, 400);
    const publishedDate = selectedArticle.published_at.toLocaleDateString('ja-JP', {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit',
    });

    return (
      <Box flexDirection="column" borderStyle="single" borderTop paddingTop={1} paddingX={1}>
        <Text bold color="green">
          {selectedArticle.title}
        </Text>
        <Box marginTop={1}>
          <Text color="gray">公開日: {publishedDate}</Text>
          {selectedArticle.author && (
            <Text color="cyan"> | 著者: {selectedArticle.author}</Text>
          )}
        </Box>
        <Box marginTop={1}>
          <Text color="yellow">URL: {selectedArticle.url}</Text>
        </Box>
        {summary && (
          <Box marginTop={1}>
            <Text>{summary}</Text>
          </Box>
        )}
        <Box marginTop={1}>
          <Text color="gray" dimColor>
            状態: {selectedArticle.is_read ? '既読' : '未読'}
            {selectedArticle.is_favorite ? ' | ★お気に入り' : ''}
          </Text>
        </Box>
      </Box>
    );
  };

  if (articles.length === 0) {
    return (
      <Box flexDirection="column" padding={1}>
        <Text bold color="blue">
          記事一覧{feedTitle && ` - ${feedTitle}`}
        </Text>
        <Box marginTop={1} justifyContent="center" alignItems="center" height={5}>
          <Text color="gray" italic>
            記事がありません
          </Text>
        </Box>
        <Box marginTop={1}>
          <Text color="yellow">ヒント: `r` でフィードを更新できます</Text>
        </Box>
      </Box>
    );
  }

  return (
    <Box flexDirection="column" height="100%" padding={1}>
      <Box>
        <Text bold color="blue">
          記事一覧{feedTitle && ` - ${feedTitle}`} ({articles.length}件)
        </Text>
      </Box>
      <Box marginTop={1} height="40%">
        <SelectableList
          items={articleItems}
          selectedIndex={selectedIndex}
          renderItem={renderArticleItem}
        />
      </Box>
      <Box marginTop={1} height="55%" flexGrow={1}>
        {renderArticleDetail()}
      </Box>
      <Box marginTop={1}>
        <Text color="gray" dimColor>
          j/k:記事選択 a:次のサイト s:前のサイト l/Enter:ブラウザで開く m:既読切替 f:お気に入り切替 r:更新 q:終了
        </Text>
      </Box>
    </Box>
  );
}
