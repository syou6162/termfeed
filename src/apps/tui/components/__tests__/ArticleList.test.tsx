import { render } from 'ink-testing-library';
import { describe, it, expect, vi } from 'vitest';
import { ArticleList } from '../ArticleList.js';
import type { Article } from '../../../../types/index.js';

// convertHtmlToTextモック
vi.mock('../../utils/html.js', () => ({
  convertHtmlToText: vi.fn((html: string) => html.replace(/<[^>]*>/g, '')),
}));

describe('ArticleList', () => {
  const createMockArticle = (
    id: number,
    options: {
      is_favorite?: boolean;
      title?: string;
      content?: string;
      author?: string;
      published_at?: Date;
    } = {}
  ): Article => ({
    id,
    feed_id: 1,
    url: `https://example.com/article${id}`,
    title: options.title || `Article ${id}`,
    content: options.content || `Content for article ${id}`,
    summary: `Summary ${id}`,
    author: options.author || undefined,
    published_at: options.published_at || new Date('2024-01-01T12:00:00Z'),
    is_read: false,
    is_favorite: options.is_favorite || false,
    thumbnail_url: undefined,
    created_at: new Date(),
    updated_at: new Date(),
  });

  const defaultProps = {
    articles: [createMockArticle(1)],
    scrollOffset: 0,
    onScrollOffsetChange: vi.fn(),
    isPinned: false,
  };

  describe('未読記事がない場合', () => {
    it('未読記事がない旨のメッセージを表示する', () => {
      const { lastFrame } = render(<ArticleList {...defaultProps} articles={[]} />);

      const output = lastFrame();
      expect(output).toContain('未読記事がありません');
      expect(output).toContain('ヒント: `r` でフィードを更新できます');
    });
  });

  describe('記事が選択されていない場合', () => {
    it('記事を選択してくださいというメッセージを表示する', () => {
      const { lastFrame } = render(<ArticleList {...defaultProps} selectedArticle={undefined} />);

      const output = lastFrame();
      expect(output).toContain('記事を選択してください');
    });

    it('未読件数を表示する', () => {
      const articles = [createMockArticle(1), createMockArticle(2)];
      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={articles} selectedArticle={undefined} />
      );

      const output = lastFrame();
      expect(output).toContain('未読2件');
    });
  });

  describe('記事詳細表示', () => {
    it('記事のタイトル、公開日、URLを表示する', () => {
      const article = createMockArticle(1, {
        title: 'Test Article Title',
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[article]} selectedArticle={article} />
      );

      const output = lastFrame();
      expect(output).toContain('Test Article Title');
      expect(output).toContain('公開日: 2024年1月15日');
      expect(output).toContain('URL: https://example.com/article1');
    });

    it('著者がある場合は著者情報を表示する', () => {
      const article = createMockArticle(1, {
        author: 'John Doe',
      });

      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[article]} selectedArticle={article} />
      );

      const output = lastFrame();
      expect(output).toContain('著者: John Doe');
    });
  });

  describe('お気に入り・ピン表示ロジック', () => {
    it('お気に入りのみの場合、公開日と同じ行に★お気に入りを表示する', () => {
      const article = createMockArticle(1, {
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={false}
        />
      );

      const output = lastFrame();

      // 公開日と同じ行にお気に入りが表示されることを検証
      expect(output).toMatch(/公開日: 2024年1月15日.*\|.*★ お気に入り/);

      // ピンは表示されない
      expect(output).not.toContain('📌 ピン');

      // 独立した行にお気に入りが表示されていないことを確認
      expect(output).not.toMatch(/★ お気に入り[\s\S]*?公開日:/);
    });

    it('ピンのみの場合、公開日と同じ行に📌ピンを表示する', () => {
      const article = createMockArticle(1, {
        is_favorite: false,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const output = lastFrame();

      // 公開日と同じ行にピンが表示されることを検証
      expect(output).toMatch(/公開日: 2024年1月15日.*\|.*📌 ピン/);

      // お気に入りは表示されない
      expect(output).not.toContain('★ お気に入り');

      // 独立した行にピンが表示されていないことを確認
      expect(output).not.toMatch(/📌 ピン[\s\S]*?公開日:/);
    });

    it('お気に入りとピン両方がある場合、公開日と同じ行に両方を表示する', () => {
      const article = createMockArticle(1, {
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const output = lastFrame();

      // 公開日と同じ行にお気に入りとピンが表示されることを検証
      expect(output).toMatch(/公開日: 2024年1月15日.*\|.*★ お気に入り.*\|.*📌 ピン/);

      // 独立した行にそれぞれが表示されていないことを確認
      expect(output).not.toMatch(/★ お気に入り[\s\S]*?公開日:/);
      expect(output).not.toMatch(/📌 ピン[\s\S]*?公開日:/);
    });

    it('お気に入りもピンもない場合、公開日行に追加情報は表示しない', () => {
      const article = createMockArticle(1, {
        is_favorite: false,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={false}
        />
      );

      const output = lastFrame();

      // お気に入りもピンも表示されない
      expect(output).not.toContain('★ お気に入り');
      expect(output).not.toContain('📌 ピン');

      // 公開日は正常に表示される
      expect(output).toContain('公開日: 2024年1月15日');
    });
  });

  describe('区切り文字の条件分岐（同一行での表示検証）', () => {
    it('著者がない場合、公開日と同一行にお気に入りが適切な区切り文字で表示される', () => {
      const article = createMockArticle(1, {
        author: undefined,
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[article]} selectedArticle={article} />
      );

      const frameOutput = lastFrame()!;
      const lines = frameOutput.split('\n');

      // 公開日とお気に入りが同一行に表示されることを検証
      const infoLine = lines.find(
        (line) => line.includes('公開日: 2024年1月15日') && line.includes('★ お気に入り')
      );

      expect(infoLine).toBeDefined();
      expect(infoLine!).toMatch(/公開日: 2024年1月15日.*\|.*★ お気に入り/);

      // 著者情報がないことを確認
      expect(infoLine!).not.toContain('著者:');
    });

    it('著者がある場合、公開日・著者・お気に入りが同一行に適切な区切り文字で表示される', () => {
      const article = createMockArticle(1, {
        author: 'John Doe',
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[article]} selectedArticle={article} />
      );

      const frameOutput = lastFrame()!;
      const lines = frameOutput.split('\n');

      // 公開日、著者、お気に入りが同一行に表示されることを検証
      const infoLine = lines.find(
        (line) =>
          line.includes('公開日: 2024年1月15日') &&
          line.includes('著者: John Doe') &&
          line.includes('★ お気に入り')
      );

      expect(infoLine).toBeDefined();
      expect(infoLine!).toMatch(/公開日: 2024年1月15日.*\|.*著者: John Doe.*\|.*★ お気に入り/);
    });

    it('著者とお気に入りがある場合、ピンも同一行に適切な区切り文字で表示される', () => {
      const article = createMockArticle(1, {
        author: 'John Doe',
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const frameOutput = lastFrame()!;
      const lines = frameOutput.split('\n');

      // すべての情報が同一行に表示されることを検証
      const infoLine = lines.find(
        (line) =>
          line.includes('公開日: 2024年1月15日') &&
          line.includes('著者: John Doe') &&
          line.includes('★ お気に入り') &&
          line.includes('📌 ピン')
      );

      expect(infoLine).toBeDefined();
      expect(infoLine!).toMatch(
        /公開日: 2024年1月15日.*\|.*著者: John Doe.*\|.*★ お気に入り.*\|.*📌 ピン/
      );
    });

    it('著者のみがある場合、公開日・著者・ピンが同一行に適切な区切り文字で表示される', () => {
      const article = createMockArticle(1, {
        author: 'John Doe',
        is_favorite: false,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const frameOutput = lastFrame()!;
      const lines = frameOutput.split('\n');

      // 公開日、著者、ピンが同一行に表示されることを検証
      const infoLine = lines.find(
        (line) =>
          line.includes('公開日: 2024年1月15日') &&
          line.includes('著者: John Doe') &&
          line.includes('📌 ピン')
      );

      expect(infoLine).toBeDefined();
      expect(infoLine!).toMatch(/公開日: 2024年1月15日.*\|.*著者: John Doe.*\|.*📌 ピン/);

      // お気に入りがないことを確認
      expect(infoLine!).not.toContain('★ お気に入り');
    });
  });

  describe('コンテンツ表示', () => {
    it('記事が選択されたときコンポーネントがレンダリングされる', () => {
      const article = createMockArticle(1, {
        content: '<p>This is test content</p>',
      });

      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[article]} selectedArticle={article} />
      );

      const output = lastFrame();
      // タイトルと基本情報が表示されることを確認
      expect(output).toContain('Article 1');
      expect(output).toContain('公開日:');
    });

    it('コンテンツがない場合でもエラーにならない', () => {
      const article = createMockArticle(1, {
        content: '',
      });

      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[article]} selectedArticle={article} />
      );

      const output = lastFrame();
      expect(output).toContain('Article 1'); // タイトルは表示される
    });
  });

  describe('レイアウト構造の検証', () => {
    it('お気に入り・ピンが独立した行として表示されていない', () => {
      const article = createMockArticle(1, {
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const frameOutput = lastFrame()!;
      const lines = frameOutput.split('\n');

      // お気に入りとピンの情報が含まれる行を探す
      const favoriteAndPinLine = lines.find(
        (line) => line.includes('★ お気に入り') && line.includes('📌 ピン')
      );

      // その行には公開日も含まれている必要がある
      expect(favoriteAndPinLine).toBeDefined();
      expect(favoriteAndPinLine!).toMatch(/公開日:/);

      // お気に入りやピンが単独の行として存在しないことを確認
      const favoriteOnlyLines = lines.filter(
        (line) => line.includes('★ お気に入り') && !line.includes('公開日:')
      );
      const pinOnlyLines = lines.filter(
        (line) => line.includes('📌 ピン') && !line.includes('公開日:')
      );

      expect(favoriteOnlyLines).toHaveLength(0);
      expect(pinOnlyLines).toHaveLength(0);
    });

    it('著者情報とお気に入り・ピンが適切な順序で同一行に表示される', () => {
      const article = createMockArticle(1, {
        author: 'John Doe',
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const output = lastFrame();

      // 正しい順序: 公開日 -> 著者 -> お気に入り -> ピン
      expect(output).toMatch(
        /公開日: 2024年1月15日.*\|.*著者: John Doe.*\|.*★ お気に入り.*\|.*📌 ピン/
      );
    });

    it('修正前の独立行レイアウトとは異なる構造になっている', () => {
      const article = createMockArticle(1, {
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      const frameOutput = lastFrame()!;

      // 修正前のような独立行パターンが存在しないことを確認
      // 例: お気に入りが単独で表示される行
      expect(frameOutput).not.toMatch(/^\s*★ お気に入り\s*$/m);
      expect(frameOutput).not.toMatch(/^\s*📌 ピン\s*$/m);

      // タイトル直後にお気に入り・ピンの独立行がないことを確認
      const lines = frameOutput.split('\n');
      const titleLineIndex = lines.findIndex((line) => line.includes('Article 1'));
      if (titleLineIndex !== -1 && titleLineIndex + 1 < lines.length) {
        const nextLine = lines[titleLineIndex + 1];
        expect(nextLine).not.toMatch(/^\s*★ お気に入り/);
        expect(nextLine).not.toMatch(/^\s*📌 ピン/);
      }
    });
  });

  describe('スクロール機能', () => {
    it('onScrollOffsetChangeが呼ばれる', () => {
      const onScrollOffsetChange = vi.fn();
      const article = createMockArticle(1);

      render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          onScrollOffsetChange={onScrollOffsetChange}
        />
      );

      // 記事選択時にスクロール位置がリセットされる
      expect(onScrollOffsetChange).toHaveBeenCalledWith(0);
    });
  });

  describe('スナップショットテスト（レイアウト破壊的変更の検出）', () => {
    it('お気に入りのみの記事レイアウト', () => {
      const article = createMockArticle(1, {
        title: 'Sample Article Title',
        author: 'John Doe',
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
        content: '<p>Sample article content with HTML tags</p>',
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={false}
        />
      );

      // お気に入りが公開日の横に表示されるレイアウトをスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-favorite-only');
    });

    it('ピンのみの記事レイアウト', () => {
      const article = createMockArticle(1, {
        title: 'Sample Article Title',
        author: 'Jane Smith',
        is_favorite: false,
        published_at: new Date('2024-01-15T10:30:00Z'),
        content: '<p>Sample article content with HTML tags</p>',
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      // ピンが公開日の横に表示されるレイアウトをスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-pin-only');
    });

    it('お気に入り＋ピン両方の記事レイアウト', () => {
      const article = createMockArticle(1, {
        title: 'Sample Article Title',
        author: 'Bob Wilson',
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
        content: '<p>Sample article content with HTML tags</p>',
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={true}
        />
      );

      // お気に入り＋ピン両方が公開日の横に表示されるレイアウトをスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-favorite-and-pin');
    });

    it('著者なし＋お気に入りの記事レイアウト', () => {
      const article = createMockArticle(1, {
        title: 'Sample Article Title',
        author: undefined,
        is_favorite: true,
        published_at: new Date('2024-01-15T10:30:00Z'),
        content: '<p>Sample article content with HTML tags</p>',
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={false}
        />
      );

      // 著者なしでお気に入りが公開日の横に表示されるレイアウトをスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-no-author-favorite');
    });

    it('通常の記事レイアウト（お気に入り・ピンなし）', () => {
      const article = createMockArticle(1, {
        title: 'Sample Article Title',
        author: 'Alice Johnson',
        is_favorite: false,
        published_at: new Date('2024-01-15T10:30:00Z'),
        content: '<p>Sample article content with HTML tags</p>',
      });

      const { lastFrame } = render(
        <ArticleList
          {...defaultProps}
          articles={[article]}
          selectedArticle={article}
          isPinned={false}
        />
      );

      // 通常の記事表示レイアウトをスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-normal');
    });

    it('未読記事なしの状態', () => {
      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={[]} selectedArticle={undefined} />
      );

      // 未読記事なしのメッセージ表示をスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-no-articles');
    });

    it('記事未選択の状態', () => {
      const articles = [createMockArticle(1), createMockArticle(2)];
      const { lastFrame } = render(
        <ArticleList {...defaultProps} articles={articles} selectedArticle={undefined} />
      );

      // 記事未選択時の表示をスナップショット
      expect(lastFrame()).toMatchSnapshot('article-layout-no-selection');
    });
  });
});
