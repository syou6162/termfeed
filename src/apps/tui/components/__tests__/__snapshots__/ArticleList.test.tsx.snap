// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > お気に入りのみの記事レイアウト > article-layout-favorite-only 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                  │
│ Sample Article Title                                                                             │
│ 公開日: 2024年1月15日 10:30 | 著者: John Doe                                                     │
│ URL: https://example.com/article1                                                                │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│ ? でヘルプ表示                                                                             1/1件 │
│                                                                                                  │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > お気に入り＋ピン両方の記事レイアウト > article-layout-favorite-and-pin 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                  │
│ Sample Article Title                                                                             │
│ 公開日: 2024年1月15日 10:30 | 著者: Bob Wilson | 📌 ピン                                         │
│ URL: https://example.com/article1                                                                │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│ ? でヘルプ表示                                                                             1/1件 │
│                                                                                                  │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > ピンのみの記事レイアウト > article-layout-pin-only 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                  │
│ Sample Article Title                                                                             │
│ 公開日: 2024年1月15日 10:30 | 著者: Jane Smith | 📌 ピン                                         │
│ URL: https://example.com/article1                                                                │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│ ? でヘルプ表示                                                                             1/1件 │
│                                                                                                  │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > 未読記事なしの状態 > article-layout-no-articles 1`] = `
"
                                        未読記事がありません
                                ヒント: \`r\` でフィードを更新できます

                                           ? でヘルプ表示
"
`;

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > 著者なし＋お気に入りの記事レイアウト > article-layout-no-author-favorite 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                  │
│ Sample Article Title                                                                             │
│ 公開日: 2024年1月15日 10:30                                                                      │
│ URL: https://example.com/article1                                                                │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│ ? でヘルプ表示                                                                             1/1件 │
│                                                                                                  │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > 記事未選択の状態 > article-layout-no-selection 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                  │
│                                      記事を選択してください                                      │
│                                                                                                  │
│                                                                                                  │
│ ? でヘルプ表示                                                                           未読2件 │
│                                                                                                  │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;

exports[`ArticleList > スナップショットテスト（レイアウト破壊的変更の検出） > 通常の記事レイアウト（お気に入り・ピンなし） > article-layout-normal 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                                  │
│ Sample Article Title                                                                             │
│ 公開日: 2024年1月15日 10:30 | 著者: Alice Johnson                                                │
│ URL: https://example.com/article1                                                                │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│                                                                                                  │
│ ? でヘルプ表示                                                                             1/1件 │
│                                                                                                  │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;
